lib/std/tar.zig

```diff
@@ -1,172 +1,1141 @@
+
+    /// File name - 0..100
+    inline fn fileName(h: V7Header) []const u8 {
+        return h.bytes[0..100];
+    }
+    /// File mode (octal) - 100..108
+    inline fn mode(h: V7Header) []const u8 {
+        return h.bytes[100..108];
+    }
+    /// Owner's numeric user ID (octal) - 108..116
+    inline fn uid(h: V7Header) []const u8 {
+        return h.bytes[108..116];
+    }
+    /// Group's numeric user ID (octal) - 116..124
+    inline fn gid(h: V7Header) []const u8 {
+        return h.bytes[116..124];
+    }
+    /// File size in bytes (octal) - 124..136
+    inline fn size(h: V7Header) []const u8 {
+        return h.bytes[124..136];
+    }
+    /// Last modification time in numeric Unix time format (octal) - 136..148
+    inline fn modTime(h: V7Header) []const u8 {
+        return h.bytes[136..148];
+    }
+    /// Checksum for header record - 148..156
+    inline fn checksum(h: V7Header) []const u8 {
+        return h.bytes[148..156];
+    }
+    /// Link indicator (file type) - 156..157
+    inline fn fileType(h: V7Header) FileType {
+        return @intToEnum(FileType, h.bytes[156]);
+    }
+    /// Name of linked file - 157..257
+    inline fn linkedFileName(h: V7Header) []const u8 {
+        return h.bytes[157..257];
     }
 
-    fn str(header: Header, start: usize, end: usize) []const u8 {
-        var i: usize = start;
-        while (i < end) : (i += 1) {
-            if (header.bytes[i] == 0) break;
+    /// Returns an (unsigned, signed) pair of checksums for the header block.
+    /// POSIX specifies a sum of the unsigned byte values, but the Sun tar used
+    /// signed byte values.
+    /// We compute and return both.
+    fn computeChecksum(h: V7Header) [2]i64 {
+        var unsigned: i64 = 0;
+        var signed: i64 = 0;
+        for (h.bytes, 0..) |_, i| {
+            const c = if (148 <= i and i < 156)
+                ' ' // Treat the checksum field itself as all spaces.
+            else
+                h.bytes[i];
+            unsigned += c;
+            signed += @bitCast(i8, c);
         }
-        return header.bytes[start..i];
+        return .{ unsigned, signed };
+    }
+
+    inline fn ustar(h: V7Header) UstarHeader {
+        return @bitCast(UstarHeader, h);
+    }
+    inline fn star(h: V7Header) StarHeader {
+        return @bitCast(StarHeader, h);
+    }
+    inline fn gnu(h: V7Header) GnuHeader {
+        return @bitCast(GnuHeader, h);
+    }
+
+    // Magics used to identify various formats.
+    const magic_gnu = "ustar ";
+    const version_gnu = " \x00";
+    const magic_ustar = "ustar\x00";
+    const version_ustar = "00";
+    const trailer_star = "tar\x00";
+
+    fn getFormat(h: V7Header) !FormatSet {
+        const value = try parseOctal(h.checksum());
+        const checksums = h.computeChecksum();
+
+        if (value != checksums[0] and value != checksums[1])
+            return fmt_unknown;
+
+        const magic = h.ustar().magic();
+        const version = h.ustar().version();
```

Comment:
You never check the version for ustar headers. You could also simplify `magic` and `version` into the same field so that they get compared all at once. They're right next to each other and are always checked together. (ie. check ustar for `ustar  \x00` and gnu for `ustar\x0000`)

Also: such an 8 byte check could easily be optimized into a single 64 bit compare.

Reply:
I agree about the 64 bit compare.  But not sure the ustar version needs to be checked here.  As is, the result is `fmt_ustar_pax` which is a bitflags of `.ustar | .pax`. This is used later in `HeaderIterator.header()` where both ustar and pax are treated the same.  

I know it looks suspicious but i think it checks out. This is also how its done [in the go implementation](https://go.dev/src/archive/tar/format.go#L186) from which portions of this implementation are ported.

Reply:
Note the use of the word 'guess' in the comments above `getFormat()` in the previous link. There are similar allusions [here](https://go.dev/src/archive/tar/common.go#L201) and [here](https://go.dev/src/archive/tar/reader.go#L162) which seem to indicate that determining the format isn't straightforward, at least in this implementation.

Reply:
all reviews have been addressed in most recent commit. i'll allow you to decide wether to 'resolve conversation'.